### PR TITLE
Catboost support  fixes #276 explain weights

### LINKED ISSUE #276: Support for CatBoost?
Hi there,

I'd love to see support for [CatBoost](https://github.com/catboost/catboost) in eli5. It has a Scikit-style API like XGBoost and LightGBM, so should be a lot of code we can copy and paste from those implementations.

I might have a dig into it myself, if no one else is already looking at it.

❤️

### PR DESCRIPTION
**Pull request by AshwinB-hat**
`explain_weights` for catboost is added with this commit along with test scenarios.

### CODE REVIEW

**Comment by lopuhin** [comment]
Thanks for PR @AshwinB-hat , it's great that you added tests and a detailed docstring right away 👍 

I noticed that you also added a lot of files from running catboost (such as files under "catboost_cached_datasets"), probably you didn't mean to commit them? I think it would be nice to exclude them from the PR.

Also a note regarding the tests, there are two things that would need to be adjusted. Tests are run in different environments, you can run them with ``tox`` (see here https://eli5.readthedocs.io/en/latest/contribute.html). Right now there are two issues:
- catboost is not installed, so the tests would not run, please add catboost requirement here: https://github.com/TeamHG-Memex/eli5/blob/55a063c77b63c47afece139e941a783b76f90fd6/tox.ini#L53
- we also run tests in enviroments when only base libraries are installed (as not everyone has catboost), the tests should be skipped in this case, you can check how this is done for lightgmb.

Will do a code review a bit later, thanks for the PR 👍

**Comment by AshwinB-hat** [comment]
@lopuhin 
Yes i will revert the commit and look into lightgbm

**Comment by AshwinB-hat** [comment]
@lopuhin 
I have made changes and commit. But 2/11 systems are still failing the tests with a catboost import error.
I have made the necessary additions in `runtests_default.sh` and `runtest_nodeps.sh`.

**Comment by AshwinB-hat** [comment]
@lopuhin  I have made the mentioned changes. 
Although one environment test is failing.

File: eli5/catboost.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
+
+from __future__ import absolute_import,division
+
+import numpy as np
```
Comment (lopuhin) [inline eli5/catboost.py]: I think the tests are failing due to missing type annotations for numpy and catboost: https://travis-ci.org/TeamHG-Memex/eli5/jobs/512862671#L777 - the way to fix this is to ignore the types here in this way: ``import numpy as np  # type: ignore``, and same for catboost import. This is something to be improved in our development process, but at the moment it is required, sorry.

File: eli5/catboost.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    elif(importance_type=="LossFunctionChange"):
+        if(isinstance(Pool,catboost.Pool)):
+            fs = catb.get_feature_importance(Pool,type=importance_type)
+        else:
+            raise ValueError(
+                'POOL datatype required or type:LossFunctionChange '
+            )
+    else:
+        raise ValueError('invalid importance_type')
+    all_features = np.array([i for i in fs],dtype=np.float32)
```
Comment (lopuhin) [inline eli5/catboost.py]: Will ``all_features = np.array(fs, dtype=np.float32)`` do the same?

File: eli5/catboost.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+def _catb_feature_importance(catb,importance_type,Pool=None):
+    if(importance_type=="PredictionValuesChange"):
+        fs = catb.get_feature_importance(type=importance_type)
+    elif(importance_type=="LossFunctionChange"):
+        if(isinstance(Pool,catboost.Pool)):
+            fs = catb.get_feature_importance(Pool,type=importance_type)
+        else:
+            raise ValueError(
+                'POOL datatype required or type:LossFunctionChange '
```
Comment (lopuhin) [inline eli5/catboost.py]: I suggest making this more specific, like ``'you need to pass `pool` (a catboost.Pool instance) for LossFunctionChange importance type'``

File: eli5/catboost.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+@explain_weights.register(catboost.CatBoost)
+@explain_weights.register(catboost.CatBoostClassifier)
+@explain_weights.register(catboost.CatBoostRegressor)
+def explain_weights_catboost(catb,
+                            vec=None,
+                            top=20,
+                            importance_type='PredictionValuesChange',
+                            prettified=True,
+                            feature_names=None,
+                            Pool=None
```
Comment (lopuhin) [inline eli5/catboost.py]: I'd suggest changing the name to lowercase ``pool``, and documenting this parameter in the docstring.

File: eli5/catboost.py [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+
+@explain_weights.register(catboost.CatBoost)
+@explain_weights.register(catboost.CatBoostClassifier)
+@explain_weights.register(catboost.CatBoostRegressor)
+def explain_weights_catboost(catb,
+                            vec=None,
+                            top=20,
+                            importance_type='PredictionValuesChange',
+                            prettified=True,
```
Comment (lopuhin) [inline eli5/catboost.py]: This argument looks unused, except for documentation (from the docs I didn't quite understand if we need it).

File: eli5/catboost.py (hunk 2 above)
Comment (AshwinB-hat) [reply eli5/catboost.py]: Yes, this is redundant. I will change it.

File: eli5/catboost.py (hunk 5 above)
Comment (AshwinB-hat) [reply eli5/catboost.py]: In the lower function. `formatter` were using vec to filter. 
I will recheck.

Comment (AshwinB-hat) [reply eli5/catboost.py]: @lopuhin  In _feature_importances\get_feature_importance_explanation, vec is given as a positional argument and I cant make it a keyword argument since it is followed by a positional argument. Changing the order breaks xgboost and lightdbm implementation. So i have just added a `vec` variable as a quick fix in the latest commit.

Comment (AshwinB-hat) [reply eli5/catboost.py]: vec is not needed for catboost as it uses its internal vectorizer when given cat_indices parameter. 
Now my doubt here was should I write some more functions insplace of get_feature_importance_explanation() for catboost or just change the ordering of parameters in the files using this function.(eg xgboost,lightgbm)

Comment (lopuhin) [reply eli5/catboost.py]: Sorry @AshwinB-hat I meant that ``prettified`` argument is unused - I should have mentioned it in the comment. Let me think about ``vec``...

File: eli5/catboost.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+        To be passed if explain_weights_catboost has importance_type set to "LossFunctionChange".
+        The catboost feature_importances_ uses the Pool datatype to calculate the parameter for the specific importance_type.
+
+    """
+    is_regression = _check_catboost_args(catb)
+    catb_feature_names = catb.feature_names_
+    coef = _catb_feature_importance(catb, importance_type=importance_type, pool=pool)
+    #vec is only added since it is a positional argument in get feature_importance_explanation and cannot 
+    #be made into a keyword argument since it is followed by a positional argument.
+    vec = None
```
Comment (lopuhin) [inline eli5/catboost.py]: Sorry for creating confusion regarding the ``vec`` variable - I think it would be nice to have it as a keyword argument, because having some vectoriser before CatBoost can still make sense - e.g. this can be vectoriser transforming words into features, so explanation would be given in terms of words and not feature names. So I suggest bringing it back into function arguments, with ``None`` default - sorry again that I was not clear here.

Comment (AshwinB-hat) [reply eli5/catboost.py]: okay will do

File: tests/test_catboost.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
+from __future__ import absolute_import
+
+import numpy as np
+from eli5.explain import explain_weights
+from .utils import format_as_all
```
[the remark this answers is not in the document]
Comment (AshwinB-hat) [reply tests/test_catboost.py]: I was not aware of the import formatting. Will do (Y)

**Comment by AshwinB-hat** [comment]
@lopuhin I have updated the PR. 
Please let me know if any other changes required.

**Comment by lopuhin** [comment]
Thanks @AshwinB-hat , PR looks great 👍 I'll check if there is anything that needs to be changed in the docs.

**Comment by lopuhin** [comment]
Hi @AshwinB-hat I checked the docs, there are a few issues, but IMO it's also fine to merge this PR and add the docs in a later PR. Here are the docs issues (btw you can check them locally if you go to ``./docs/`` folder, install requirements and run ``make html``):
- we'd like to have a section about Catboost under "Supported Libraries", this should go into ``source/libraries/catboost.rst``, see lightgbm docs for an example
- we'd like to add Catboost to supported features in ``source/overview.rst``
- we'd like to build API docs for catboost, this should go into ``source/autodocs/catboost.rst`` (see lightgbm docs for an example), and include it into ``source/autodocs/index.rst``
- there are a few warnings produced when building catboost docs (they appear once you include them as per above item), would be great to fix them.

@AshwinB-hat please tell if you'd like to finish the docs in this PR (this can be done later as well), or proceed with the merge?

**Comment by AshwinB-hat** [comment]
@lopuhin I prefer I make a different PR for the Catboost docs. 

Currently I am investigating the SHAP paper and writing my proposal which I might run by @lopuhin @kmike  in a day or two for feedback. 

I will look into docs following the submission of my GSOC proposal and start working on the catboost explain_prediction as well.

That is my plan of action as of now.

**Comment by lopuhin** [comment]
@AshwinB-hat great, that's a solid plan 👍 

@kmike would you like to check this PR?

**Comment by kmike** [comment]
Hey! Could you please add docs in the same pull request, to make it complete? I think that's good to have all features tested and documented.

**Comment by AshwinB-hat** [comment]
@kmike 
Yes, will complete the documentation.

**Comment by AshwinB-hat** [comment]
@kmike @lopuhin 
I have added the docs and resolved the warnings. They were arising due to the tutorials not being added in toc tree. 

I am done with my exams and other commitments hence the delay.  I would be interested on working on the next step that is the prediction.

Also Let me know if any improvement or changes.

**Comment by AshwinB-hat** [comment]
hi @lopuhin,
You must be flooded with notifications. Just making sure you have seen the above comment. 
If there are more changes to be done before the merge, please let me know. 
It will be mean a lot to me if this change was merged.. as it is my first contribution.

**Comment by lopuhin** [comment]
Hi @AshwinB-hat , many thanks for adding the docs and resolving comments, I gave a quick look and it looks great, will check in more detail tomorrow morning.

File: eli5/__init__.py [hunk 8]
```diff
@@ -73,6 +72,12 @@
 except ImportError:
     # lightgbm is not available
     pass
-except OSError:
```
Comment (ivanprado) [inline eli5/__init__.py]: I guess this check was removed by error. Could you put it back?

**Comment by ivanprado** [comment]
Hi @AshwinB-hat! I'll have a look at the PR so that we can have it merged soon :-). 

Could you please update the README file to include the new feature? (https://github.com/AshwinB-hat/eli5/blob/master/README.rst)

**Comment by AshwinB-hat** [comment]
@ivanprado Yes! will do it right away!

File: eli5/__init__.py (hunk 8 above)
Comment (ivanprado) [reply eli5/__init__.py]: I have seen that you added this `except` to all import trials in the file. Not sure this is right. Probably `lightgbm` included this `except` because in case of bad installation is failing improperly, but for the rest, I think is better to keep them as they were. Could you please leave `OSError` check only for `lightgbm` as it was before?

**Comment by ivanprado** [comment]
Hi @AshwinB-hat !

I see some formatting issues in the documentation. That's what I see: 
image
Could you please have a look? The best way is by building documentation by yourself. Here you have the instructions: https://github.com/TeamHG-Memex/eli5/tree/master/docs

File: docs/source/autodocs/catboost.rst [hunk 9]
```diff
@@ -0,0 +1,11 @@
+eli5.catboost
+=============
+
+eli5 has CatBoost_ support - :func:`eli5.explain_weights`
+shows feature importances, 
+The function work for CatBoost, CatBoostClassifier and CatBoostRegressor.
```
Comment (ivanprado) [inline docs/source/autodocs/catboost.rst]:
```suggestion
The function works for CatBoost, CatBoostClassifier and CatBoostRegressor.
```

File: docs/source/libraries/catboost.rst [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+:func:`eli5.explain_weights` uses feature importances. Additional
+arguments for CatBoostClassifier and CatBoostRegressor:
+
+* ``importance_type`` is a way to get feature importance. Possible values are:
+
+  - 'PredictionValuesChange' - The individual importance values for each of the input features.
+          (default)
+  - 'LossFunctionChange' - The individual importance values for each of the input features for ranking metrics (requires training data to be passed  or a similar dataset with Pool)
+
+* ``pool`` the catboost.Pool datatype . To be passed if explain_weights_catboost has importance_type set to "LossFunctionChange".The catboost feature_importances uses the Pool datatype to calculate the parameter for the specific importance_type.
```
Comment (ivanprado) [inline docs/source/libraries/catboost.rst]:
```suggestion
* ``pool`` the catboost.Pool datatype . To be passed if explain_weights_catboost has importance_type set to "LossFunctionChange". The catboost feature_importances uses the Pool datatype to calculate the parameter for the specific importance_type.
```

Comment (ivanprado) [reply docs/source/libraries/catboost.rst]: In general, I think that arguments and classes should be wrapped with ``  and variable or arguments values be wrapped by '

**Comment by ivanprado** [comment]
https://github.com/TeamHG-Memex/eli5/pull/313 has been merged in master, and that should fix the build. Could you please merge back master with your fork? That should make Travis happy :-)

File: docs/source/autodocs/catboost.rst (hunk 9 above)
Comment (AshwinB-hat) [reply docs/source/autodocs/catboost.rst]: roger!

File: docs/source/libraries/catboost.rst (hunk 10 above)
Comment (AshwinB-hat) [reply docs/source/libraries/catboost.rst]: Yes! makes sense. Fixing the issues.

**Comment by AshwinB-hat** [comment]
@ivanprado I made the suggested changes and some. Looks fine on the local build (thanks for the tip).
Travis seems relatively happier since before. :)

**Comment by ivanprado** [comment]
Nice @AshwinB-hat, I'll have a look to the changes! I see the following error in Travis:
`/home/travis/build/TeamHG-Memex/eli5/docs/source/autodocs/catboost.rst:4:Inline literal start-string without end-string.` 
Could you please have a look?

File: docs/source/autodocs/catboost.rst [hunk 11]
```diff
@@ -3,7 +3,7 @@ eli5.catboost
 
 eli5 has CatBoost_ support - :func:`eli5.explain_weights`
 shows feature importances, 
-The function work for CatBoost, CatBoostClassifier and CatBoostRegressor.
+The function works for ``CatBoost``, ``CatBoostClassifier and CatBoostRegressor.
```
Comment (ivanprado) [inline docs/source/autodocs/catboost.rst]: Probably this line is causing the error.

**Comment by ivanprado** [comment]
Ok, I had a look, we are very close! Remaining:
- [ ] OSError handling on `__init__.py` https://github.com/TeamHG-Memex/eli5/pull/303#discussion_r287830809
- [ ] Making Travis happy again :-)
- [ ] Adding test cases for `catboost.py` lines 71 and 75 if possible. See Codecov report here (lines in red): https://codecov.io/gh/TeamHG-Memex/eli5/pull/303/diff?src=pr&el=tree#D13-75 . Codecov is configured with the following policy: new code should increase test coverage, never decrease it. Let's try to follow it also in this case :-)

**Comment by AshwinB-hat** [comment]
@ivanprado Guess its good news!

File: eli5/__init__.py [hunk 12]
```diff
@@ -74,5 +69,12 @@
     # lightgbm is not available
     pass
 except OSError:
-    # improperly installed lightgbm
```
Comment (ivanprado) [inline eli5/__init__.py]: Could you keep this comment? I think it is informative.

File: eli5/__init__.py [hunk 13]
```diff
@@ -74,5 +69,12 @@
...
-    pass
\ No newline at end of file
+    pass
+
+try:
+    from .catboost import (
+        explain_weights_catboost
+    )
+except ImportError:
+    #catboost is not available
```
Comment (ivanprado) [inline eli5/__init__.py]:
```suggestion
    # catboost is not available
```

**Comment by ivanprado** [comment]
@AshwinB-hat congratulations!. Your first contribution is now in master :-)

**Comment by kmike** [comment]
🎉

**Comment by AshwinB-hat** [comment]
@lopuhin @ivanprado @kmike Awesome. Thanks for the guidance leading up till this. :D
